controllers/idle_handlers.go

```diff
@@ -18,17 +18,76 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"os"
 
 	lcav1alpha1 "github.com/openshift-kni/lifecycle-agent/api/v1alpha1"
+	"github.com/openshift-kni/lifecycle-agent/controllers/utils"
+	ostree "github.com/openshift-kni/lifecycle-agent/internal/ostree"
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+func cleanupUnbootedStateroot(stateroot string) error {
+	status, err := ostree.QueryStatus()
+	if err != nil {
+		return err
+	}
+
+	undeployIndices := make([]int, 0)
+	for index, deployment := range status.Deployments {
+		if deployment.OSName != stateroot {
+			continue
+		}
+		// make sure none of the deployments in the stateroot is booted
+		if deployment.Booted {
+			return fmt.Errorf("failed abort: deployment %d in stateroot %s is booted", index, stateroot)
+		}
+		undeployIndices = append(undeployIndices, index)
+	}
+
+	// since undeploy shifts the order, undeploy in the reverse order
+	for i := len(undeployIndices) - 1; i >= 0; i-- {
```

Comment:
Since you're collect the index in `undeployIndices` and skipping when it's new stateroot (L45)...it be should be safe undeploy regardless of the direction of the loop?